tensorboard/data/proto/data_provider.proto

```diff
@@ -0,0 +1,295 @@
+  // Only match tags with exactly one of these names. In particular, if this
+  // list is empty, no tags match.
+  repeated string tags = 1;
+}
+
+message ScalarMetadata {
+  // Largest step value of any datum in this time series. Non-negative.
+  int64 max_step = 1;
+  // Largest wall time of any datum in this time series.
+  double max_wall_time = 2;
+  // Atemporal summary metadata for this time series.
+  tensorboard.SummaryMetadata summary_metadata = 3;
+}
+
+message ReadScalarsRequest {
+  // ID of experiment in which to query data.
+  string experiment_id = 1;
+  // Required filter for plugin name. If omitted, an empty message is implied.
+  PluginFilter plugin_filter = 2;
+  // Optional filter for time series. If omitted, all time series match.
+  RunTagFilter run_tag_filter = 3;
+  // Required downsampling specification describing how many points to return
+  // per time series. If omitted, an empty message is implied, which means that
+  // no points will be returned.
+  Downsample downsample = 4;
+}
+
+message Downsample {
+  // Maximum number of points to return. Should be non-negative. Zero means
+  // zero.
+  int64 num_points = 1;
+}
+
+message ReadScalarsResponse {
+  repeated RunEntry runs = 1;
+  message RunEntry {
+    string run_name = 1;
+    repeated TagEntry tags = 2;
+  }
+  message TagEntry {
+    string tag_name = 1;
+    ScalarData data = 2;
+  }
+}
+
+// A column-major sequence of scalar points. Arrays `step`, `wall_time`, and
+// `value` have the same lengths.
+message ScalarData {
+  repeated int64 step = 1 [packed = true];
+  repeated double wall_time = 2 [packed = true];
+  repeated double value = 3 [packed = true];
+}
+
+message ListTensorsRequest {
+  // ID of experiment in which to query data.
+  string experiment_id = 1;
+  // Required filter for plugin name. If omitted, an empty message is implied.
+  PluginFilter plugin_filter = 2;
+  // Optional filter for time series. If omitted, all time series match.
+  RunTagFilter run_tag_filter = 3;
+}
+
+message ListTensorsResponse {
+  repeated RunEntry runs = 1;
+  message RunEntry {
+    string run_name = 1;
+    repeated TagEntry tags = 2;
+  }
+  message TagEntry {
+    string tag_name = 1;
+    TensorMetadata time_series = 2;
+  }
+}
+
+message TensorMetadata {
+  // Largest step value of any datum in this time series. Non-negative.
+  int64 max_step = 1;
+  // Largest wall time of any datum in this time series.
+  double max_wall_time = 2;
+  // Atemporal summary metadata for this time series.
+  tensorboard.SummaryMetadata summary_metadata = 3;
+}
+
+message ReadTensorsRequest {
+  // ID of experiment in which to query data.
+  string experiment_id = 1;
+  // Required filter for plugin name. If omitted, an empty message is implied.
+  PluginFilter plugin_filter = 2;
+  // Optional filter for time series. If omitted, all time series match.
+  RunTagFilter run_tag_filter = 3;
+  // Required downsampling specification describing how many points to return
```

Comment:
Again, it's a bit confusing to me to describe this as "Required" and then specify what happens if it's omitted to be something other than returning an error that the request is invalid.